amlip_cpp/src/cpp/types/AmlipIdDataType.cpp

```diff
@@ -0,0 +1,310 @@
+
+AmlipIdDataType::AmlipIdDataType()
+{
+    memset(&name_, 0, (NAME_SIZE) * 1);
+    memset(&rand_id_, 0, (RAND_SIZE) * 1);
+}
+
+AmlipIdDataType::AmlipIdDataType(const std::string& name)
+{
+    name_ = str_name_to_array_(name);
+    rand_id_ = random_id_();
+}
+
+AmlipIdDataType::AmlipIdDataType(const char* name)
+    : AmlipIdDataType(std::string(name))
+{
+}
+
+AmlipIdDataType::AmlipIdDataType(
+    std::array<uint8_t, NAME_SIZE>& name,
+    std::array<uint8_t, RAND_SIZE>& rand_id)
+{
+    name_ = name;
+    rand_id_ = rand_id;
+}
+
+AmlipIdDataType::AmlipIdDataType(
+    std::array<uint8_t, NAME_SIZE>&& name,
+    std::array<uint8_t, RAND_SIZE>&& rand_id)
+{
+    name_ = std::move(name);
+    rand_id_ = std::move(rand_id);
+}
+
+AmlipIdDataType::~AmlipIdDataType()
+{
+}
+
+AmlipIdDataType::AmlipIdDataType(
+    const AmlipIdDataType& x)
+{
+    name_ = x.name_;
+    rand_id_ = x.rand_id_;
+}
+
+AmlipIdDataType::AmlipIdDataType(
+    AmlipIdDataType&& x)
+{
+    name_ = std::move(x.name_);
+    rand_id_ = std::move(x.rand_id_);
+}
+
+AmlipIdDataType& AmlipIdDataType::operator =(
+    const AmlipIdDataType& x)
+{
+    name_ = x.name_;
+    rand_id_ = x.rand_id_;
+
+    return *this;
+}
+
+AmlipIdDataType& AmlipIdDataType::operator =(
+    AmlipIdDataType&& x)
+{
+    name_ = std::move(x.name_);
+    rand_id_ = std::move(x.rand_id_);
+
+    return *this;
+}
+
+bool AmlipIdDataType::operator ==(
+    const AmlipIdDataType& x) const
+{
+    return (name_ == x.name_ && rand_id_ == x.rand_id_);
+}
+
+bool AmlipIdDataType::operator !=(
+    const AmlipIdDataType& x) const
+{
+    return !(*this == x);
+}
+
+std::string AmlipIdDataType::name() const
+{
+    std::string name = std::string(std::begin(name_), std::end(name_));
+    uint8_t null_idx = name.find_first_of('\0');
+    if (null_idx < name.size())
+    {
+        return name.erase(null_idx);
+    }
+    else
+    {
+        return name;
+    }
+}
+
+void AmlipIdDataType::name(const std::array<uint8_t, NAME_SIZE>& name)
+{
+    name_ = name;
+}
+
+const std::array<uint8_t, NAME_SIZE>& AmlipIdDataType::base64_name() const
+{
+    return name_;
+}
+
+const std::array<uint8_t, RAND_SIZE>& AmlipIdDataType::id() const
+{
+    return rand_id_;
+}
+
+void AmlipIdDataType::id(const std::array<uint8_t, RAND_SIZE>& id)
+{
+    rand_id_ = id;
+}
+
+const char* AmlipIdDataType::type_name()
+{
+    return "AMLIP-ID";
+}
+
+bool AmlipIdDataType::is_defined() const noexcept
+{
+    return (*this) != UNDEFINED_ID_;
+}
+
+AmlipIdDataType AmlipIdDataType::new_unique_id()
+{
+    return AmlipIdDataType(random_name_(), random_id_());
+}
+
+AmlipIdDataType AmlipIdDataType::new_unique_id(const std::string& name)
+{
+    AmlipIdDataType new_id = new_unique_id();
+    new_id.name(str_name_to_array_(name));
+
+    return new_id;
+}
+
+AmlipIdDataType AmlipIdDataType::new_unique_id(const char* name)
+{
+    return new_unique_id(std::string(name));
```

Comment:
*In general*
It is normally more convenient to transform a char* into a string and not otherwise. Cast from char to string creates a new object, and the other way around it just passes a pointer.